### PR TITLE
[Azure] Add Microsoft Graph Activity Logs datastream

### PR DESCRIPTION
**Pull request by kcreddy**
## Proposed commit message
- Add Microsoft Graph Activity Logs datastream.

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

`elastic-package build && elastic-package stack up -d -v && eval "$(elastic-package stack shellinit)" && elastic-package test pipeline --generate  -v `

## Related issues

- Closes https://github.com/elastic/integrations/issues/8555

## Screenshots

graph_activity_logs

### CODE REVIEW

File: packages/azure/data_stream/graphactivitylogs/agent/stream/azure-eventhub.yml.hbs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+storage_account_key: {{storage_account_key}}
+{{/if}}
+{{#if resource_manager_endpoint}}
+resource_manager_endpoint: {{resource_manager_endpoint}}
+{{/if}}
+tags:
+{{#if preserve_original_event}}
+  - preserve_original_event
+{{/if}}
+{{#each tags as |tag i|}}
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/agent/stream/azure-eventhub.yml.hbs:30]:
```suggestion
{{#each tags as |tag|}}
```

File: packages/azure/data_stream/graphactivitylogs/agent/stream/azure-eventhub.yml.hbs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+processors:
+{{processors}}
+{{/if}}
+sanitize_options:
+{{#if sanitize_newlines}}
+  - NEW_LINES
+{{/if}}
+{{#if sanitize_singlequotes}}
+  - SINGLE_QUOTES
+{{/if}}
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/agent/stream/azure-eventhub.yml.hbs:46]: Final new line.

File: packages/azure/data_stream/graphactivitylogs/agent/stream/azure-eventhub.yml.hbs [hunk 3]
```diff
@@ -0,0 +1,46 @@
+{{#if connection_string}}
+connection_string: {{connection_string}}
+{{/if}}
+{{#if storage_account_container }}
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/agent/stream/azure-eventhub.yml.hbs:4]:
```suggestion
{{#if storage_account_container}}
```

File: packages/azure/data_stream/graphactivitylogs/agent/stream/log.yml.hbs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+paths:
+  {{#each paths as |path i|}}
+- {{path}}
+  {{/each}}
+exclude_files: [".gz$"]
+tags:
+{{#if preserve_original_event}}
+  - preserve_original_event
+{{/if}}
+{{#each tags as |tag i|}}
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/agent/stream/log.yml.hbs:10]:
```suggestion
{{#each tags as |tag|}}
```

File: packages/azure/data_stream/graphactivitylogs/fields/fields.yml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+          description: |
+            The client request identifier when sent. If no client request identifier is sent, the value will be equal to the operation identifier.
+        - name: identity_provider
+          type: keyword
+          description: |
+            The identity provider that authenticated the subject of the token.
+        - name: is_billable
+          type: boolean
+          description: |
+            The record size in bytes.
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/fields/fields.yml:56]: This doesn't look right.

File: packages/azure/data_stream/graphactivitylogs/fields/fields.yml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+          description: |
+            The identifier of the servicePrincipal making the request.
+        - name: sign_in_activity_id
+          type: keyword
+          description: |
+            The identifier representing the sign-in activitys.
+        - name: source_system
+          type: keyword
+          description: |
+            The type of agent the event was collected by. For example, OpsManager for Windows agent, either direct connect or Operations Manager, Linux for all Linux agents, or Azure for Azure Diagnostics
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/fields/fields.yml:84]:
```suggestion
            The type of agent the event was collected by. For example, OpsManager for Windows agent, either direct connect or Operations Manager, Linux for all Linux agents, or Azure for Azure Diagnostics.
```

File: packages/azure/data_stream/graphactivitylogs/fields/package-fields.yml [hunk 7]
```diff
@@ -0,0 +1,42 @@
+- name: azure
+  type: group
+  fields:
+    - name: subscription_id
+      type: keyword
+      description: |
+        Azure subscription ID
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/fields/package-fields.yml:7]: Final full-stops throughout.

File: packages/azure/data_stream/graphactivitylogs/manifest.yml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        default: false
+      - name: storage_account_container
+        type: text
+        title: Storage Account Container
+        multi: false
+        required: false
+        show_user: false
+        description: >
+          The storage account container where the integration stores the checkpoint data for the consumer group. It is an advanced option to use with extreme care. You MUST use a dedicated storage account container for each Azure log type (activity, sign-in, audit logs, and others). DO NOT REUSE the same container name for more than one Azure log type. See [Container Names](https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names) for details on naming rules from Microsoft. The integration generates a default container name if not specified.
+
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/manifest.yml:26]:
```suggestion
        description: >-
          The storage account container where the integration stores the checkpoint data for the consumer group. It is an advanced option to use with extreme care. You MUST use a dedicated storage account container for each Azure log type (activity, sign-in, audit logs, and others). DO NOT REUSE the same container name for more than one Azure log type. See [Container Names](https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names) for details on naming rules from Microsoft. The integration generates a default container name if not specified.
```
(not sure if GH is correctly including the deletion of the final blank line)

File: packages/azure/_dev/build/docs/graphactivitylogs.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
+# Microsoft Graph Activity Logs
+
+Microsoft Graph Activity Logs provide an audit trail of all HTTP requests that the Microsoft Graph service received and processed for a tenant. Microsoft Graph Activity Logs gives full visibility into all transactions made by applications and other API clients that you have consented to in the tenant. Refer to [Microsoft Graph Activity Common Usecases](https://learn.microsoft.com/en-us/graph/microsoft-graph-activity-logs-overview#common-use-cases-for-microsoft-graph-activity-logs) for more use cases.
```
Comment (efd6) [inline packages/azure/_dev/build/docs/graphactivitylogs.md:3]:
```suggestion
Microsoft Graph Activity Logs provide an audit trail of all HTTP requests that the Microsoft Graph service has received and processed for a tenant. Microsoft Graph Activity Logs gives full visibility into all transactions made by applications and other API clients that you have consented to in the tenant. Refer to [Microsoft Graph Activity Common Usecases](https://learn.microsoft.com/en-us/graph/microsoft-graph-activity-logs-overview#common-use-cases-for-microsoft-graph-activity-logs) for more use cases.
```

File: packages/azure/_dev/build/docs/graphactivitylogs.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+Tenant administrators can configure the collection and storage destinations of Microsoft Graph Activity Logs through Diagnostic Setting in the Entra Portal. This integrtaion uses Azure Event Hubs destination to stream Microsoft Graph Activity Logs to Elastic.
```
Comment (efd6) [inline packages/azure/_dev/build/docs/graphactivitylogs.md:5]:
```suggestion
Tenant administrators can configure the collection and storage destinations of Microsoft Graph Activity Logs through Diagnostic Setting in the Entra Portal. This integration uses Azure Event Hubs destination to stream Microsoft Graph Activity Logs to Elastic.
```

File: packages/azure/_dev/build/docs/graphactivitylogs.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+- Activities of multi-tenant applications belonging to another tenant are not available.
+- In few rare cases, events might take up to 2 hours to be delivered to Event Hubs.
+Refer to [Microsoft Graph Activity Limitations](https://learn.microsoft.com/en-us/graph/microsoft-graph-activity-logs-overview#limitations) for more information.
+
+## Settings
+
+`eventhub` :
+  _string_
+It is a fully managed, real-time data ingestion service. Elastic recommends using only letters, numbers, and the hyphen (-) character for Event Hub names to maximize compatibility. You can use existing Event Hubs having underscores (_) in the Event Hub name; in this case, the integration will replace underscores with hyphens (-) when it uses the Event Hub name to create dependent Azure resources behind the scenes (e.g., the storage account container to store Event Hub consumer offsets). Elastic also recommends using a separate event hub for each log type as the field mappings of each log type differ.
```
Comment (efd6) [inline packages/azure/_dev/build/docs/graphactivitylogs.md:30]:
```suggestion
It is a fully managed, real-time data ingestion service. Elastic recommends using only letters, numbers, and the hyphen (-) character for Event Hub names to maximize compatibility. You _can_ use existing Event Hubs having underscores (_) in the Event Hub name; in this case, the integration will replace underscores with hyphens (-) when it uses the Event Hub name to create dependent Azure resources behind the scenes (e.g., the storage account container to store Event Hub consumer offsets). Elastic also recommends using a separate event hub for each log type as the field mappings of each log type differ.
```

File: packages/azure/_dev/build/docs/graphactivitylogs.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+_string_
+The storage account key, this key will be used to authorize access to data in your storage account.
+
+`storage_account_container` :
+_string_
+The storage account container where the integration stores the checkpoint data for the consumer group. It is an advanced option to use with extreme care. You MUST use a dedicated storage account container for each Azure log type (activity, sign-in, audit logs, and others). DO NOT REUSE the same container name for more than one Azure log type. See [Container Names](https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names) for details on naming rules from Microsoft. The integration generates a default container name if not specified. 
+
+`resource_manager_endpoint` :
+_string_
+Optional, by default we are using the azure public environment, to override, users can provide a specific resource manager endpoint in order to use a different azure environment.
```
Comment (efd6) [inline packages/azure/_dev/build/docs/graphactivitylogs.md:58]:
```suggestion
Optional, by default we are using the azure public environment, to override, users can provide a specific resource manager endpoint in order to use a different Azure environment.
```
?

File: packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/azure-shared-pipeline.yml [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+      field: azure.resource_id
+      target_field: azure.resource.id
+      ignore_missing: true
+  - lowercase:
+      field: event.outcome
+      ignore_missing: true
+on_failure:
+  - set:
+      field: error.message
+      value: '{{ _ingest.on_failure_message }}'
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/azure-shared-pipeline.yml:91]: Final new line.

File: packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml [hunk 14]
```diff
@@ -0,0 +1,472 @@
+---
+description: Pipeline for parsing azure graph activity logs.
+processors:
+  - set:
+      field: ecs.version
+      value: '8.0.0'
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml:6]: Bump?

File: packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+      field: azure
+      target_field: azure-eventhub
+      ignore_missing: true
+      tag: rename-azure
+  - script:
+      source: ctx.message = ctx.message.replace(params.empty_field_name, '')
+      params:
+        empty_field_name: '"":"",'
+      ignore_failure: true
+      tag: script-message-emptyfields
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml:26]: This is a surprising order of fields.

File: packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml [hunk 16]
```diff
@@ -0,0 +1,472 @@
...
+      tag: rename-duration_ms
+  - rename:
+      field: azure.graphactivitylogs.properties.duration_ms
+      target_field: event.duration
+      if: ctx.azure?.graphactivitylogs?.properties?.durationMs != null && ctx.event?.duration == null
+      tag: rename-properties-durationMs
+  - script:
+      lang: painless
+      source: if (ctx.event.duration!= null) {ctx.event.duration = ctx.event.duration
+        * params.param_nano;}
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml:179]:
```suggestion
      source: |
        if (ctx.event.duration!= null) {
          ctx.event.duration = ctx.event.duration * params.param_nano;
        }
```

File: packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml [hunk 17]
```diff
@@ -0,0 +1,472 @@
...
+      params:
+        param_nano: 1000000
```
Comment (efd6) [inline packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml:181]: Can this not just be a literal const?

File: packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml (hunk 14 above)
Comment (kcreddy) [reply packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml:6]: I wanted to, but the package has dependency with `8.0.0` and multiple datastreams might be impacted.

https://github.com/elastic/integrations/blob/main/packages/azure/_dev/build/build.yml#L1-L3

File: packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml (hunk 15 above)
Comment (kcreddy) [reply packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml:26]: It is. I took a base ingest pipeline template and modified for this usecase. 
This processor is present in most of the package's datastreams. I wonder if its even required here. Might as well remove it.

File: packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml (hunk 17 above)
Comment (kcreddy) [reply packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml]: Updated now.

File: packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml (hunk 15 above)
Comment (kcreddy) [reply packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml]: Removed this processor in the new commit.

File: packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml (hunk 14 above)
Comment (efd6) [reply packages/azure/data_stream/graphactivitylogs/elasticsearch/ingest_pipeline/default.yml:6]: OK

**Review by efd6 — APPROVED** [review]

**Comment by kcreddy** [comment]
> @jamiehynds I don't have any feedback at this time, but I'm looking forward to testing out this integration and I may have some feedback after using it with some live data.

@aarju This feature is now available. Please feel free to test and provide feedback. Thanks 😄

**Pull request merged**